kola/tests: add test for restarting containerd

ajeddeloh opened the pull request:
Add test to kill containerd and ensure it restarts. Refactor docker info
logic to be common between docker info test and this test.

Depends on: https://github.com/coreos/coreos-overlay/pull/2703

euank commented on kola/tests/docker/docker.go:
```diff
@@ -459,17 +470,75 @@ func dockerUserNoCaps(c cluster.TestCluster) {
+		c.Fatal(err)
+	}
+
+	if testContainerdUp(c) {
```
This seems like it would be racy since systemd could have already restarted it, right?

Perhaps checking if the pid changes would be a more robust way to make sure this test isn't giving a false positive?

euank commented on kola/tests/docker/docker.go:
```diff
@@ -459,17 +470,75 @@ func dockerUserNoCaps(c cluster.TestCluster) {
+	}
+
+	// verify it came back
+	if !testContainerdUp(c) {
```
This test should perhaps also check that `systemctl is-active containerd` succeeds since it's testing that containerd was restarted by systemd.

euank commented on kola/tests/docker/docker.go:
```diff
@@ -459,17 +470,75 @@ func dockerUserNoCaps(c cluster.TestCluster) {
+	}
+
+	// kill it and verify it went down
+	pid, err := m.SSH("pgrep containerd")
```
`systemctl show containerd -p MainPID --value` is arguably more robust.

I'm personally always wary of `pgrep`.

ajeddeloh commented:
Added sugguestions, using systemctl info again instead of is-active (kill two birds with one stone). Also added a delay between killing and checking if it's up to avoid races.

bgilbert commented on kola/tests/docker/docker.go:
```diff
@@ -170,6 +170,17 @@ systemd:
+  units:
+   - name: docker.service
+     enable: true`),
+	})
```
You'll want to set a `MinVersion` to avoid uninteresting failures.

ajeddeloh replied:
Since the change isn't in any channel yet, what should it be set to?

bgilbert replied:
The next-next alpha, which will be 1520.  Of course you'll need to disable that field when testing.

ajeddeloh replied:
Fixed

bgilbert commented on kola/tests/docker/docker.go:
```diff
@@ -459,17 +470,86 @@ func dockerUserNoCaps(c cluster.TestCluster) {
+}
+
+func testContainerdUp(c cluster.TestCluster) bool {
+	type containerdInfo struct {
```
Let's move the struct to file scope, rather than redefining parts of it inside different tests.

bgilbert commented on kola/tests/docker/docker.go:
```diff
@@ -459,17 +470,86 @@ func dockerUserNoCaps(c cluster.TestCluster) {
+	}
+
+	// give it a chance to restart
+	if _, err = m.SSH("sleep 2"); err != nil {
```
This will cause test flakes on slow nodes.  It'd be better to poll the unit status until it's stable -- either running or stopped.

bgilbert commented on kola/tests/docker/docker.go:
```diff
@@ -459,17 +470,86 @@ func dockerUserNoCaps(c cluster.TestCluster) {
+		c.Fatalf("Containerd did not start")
+	}
+
+	pid, err := m.SSH("systemctl show containerd -p MainPID")
```
Why not `--value`?

euank commented on kola/tests/docker/docker.go:
```diff
@@ -459,39 +492,98 @@ func dockerUserNoCaps(c cluster.TestCluster) {
+	return info.ContainerdCommit.ID == info.ContainerdCommit.Expected
+}
+
+func unmarshalDockerInfo(m platform.Machine) (simplifiedDockerInfo, error) {
```
this is a little bikesheddy, but I think this method name is misleading. `unmarshal` to me makes it sound like it's idempotent / not doing network traffic.

Maybe `retrieveDockerInfo` or `queryDockerInfo`?

euank commented on kola/tests/docker/docker.go:
```diff
@@ -459,39 +492,98 @@ func dockerUserNoCaps(c cluster.TestCluster) {
+
+}
+
+func testContainerdUp(c cluster.TestCluster) bool {
```
More bikeshedding; this doesn't have to return a bool now that there's no attempt to observe containerd not being up.

It can just `c.Fatalf` itself if the commits don't match. Arguably `assert` is clearer than `test` when the helper is doing that, but I know I already started the pattern of using `test` in this file.

ajeddeloh commented:
Fixed @euank's nits. I don't have write access
